crates/proof_of_stake/src/lib.rs

```diff
@@ -2913,3 +2946,72 @@ where
 
     Ok(())
 }
+
+fn add_delegation_target<S>(
+    storage: &mut S,
+    delegator: &Address,
+    validator: &Address,
+    epoch: Epoch,
+    _current_epoch: Epoch,
+) -> namada_storage::Result<()>
+where
+    S: StorageRead + StorageWrite,
+{
+    let bond_holders = delegation_targets_handle(delegator);
+    if let Some(delegations) = bond_holders.get(storage, validator)?.as_mut() {
+        let (start, end) = delegations.last_range;
+        if let Some(end) = end {
+            // Add the `last_range` pair to the `prev_ranges` and make a new
+            // `last_range`
+            delegations.prev_ranges.push((start, end));
+            delegations.last_range = (epoch, None);
+            bond_holders.insert(
+                storage,
+                validator.clone(),
+                delegations.clone(),
+            )?;
+        } else {
+            // do nothing since the last bond is still active
+        }
+    } else {
+        // Make a new delegation to this source-validator pair
+        let first_delegation = DelegationEpochs {
+            prev_ranges: vec![],
+            last_range: (epoch, None),
+        };
+        bond_holders.insert(storage, validator.clone(), first_delegation)?;
+    }
+
+    // Todo: possibly update the data by pruning old pairs in
+    // `delegations.prev_ranges`??
+
+    Ok(())
+}
+
+fn remove_delegation_target<S>(
+    storage: &mut S,
+    delegator: &Address,
+    validator: &Address,
+    epoch: Epoch,
+    _current_epoch: Epoch,
+) -> namada_storage::Result<()>
+where
+    S: StorageRead + StorageWrite,
+{
+    let validators = delegation_targets_handle(delegator);
+    if let Some(delegation) = validators.get(storage, validator)?.as_mut() {
+        debug_assert!(
+            delegation.last_range.1.is_none(),
+            "End epoch should be None since we are removing the delegation
+              right now!!"
+        );
+        delegation.last_range.1 = Some(epoch);
```

Comment:
```suggestion
        let (_start, end) = &mut delegation.last_range;
        debug_assert!(
            end.is_none(),
            "End epoch should be None since we are removing the delegation
              right now!!"
        );
        *end = Some(epoch);
```

Reply:
thx! included in [c7574fd](https://github.com/anoma/namada/pull/3043/commits/c7574fd353dcd2da1cb21f2bd2efd4eff1bb38c5)